plugins/inputs/cpufreq/cpufreq.go

```diff
@@ -0,0 +1,157 @@
+// +build linux
+
+package cpufreq
+
+import (
+	"fmt"
+	"io"
+	"os"
+	"path"
+	"path/filepath"
+	"strconv"
+	"strings"
+
+	"github.com/influxdata/telegraf"
+	"github.com/influxdata/telegraf/plugins/inputs"
+)
+
+const defaultHostSys = "/sys"
+
+type CPUFreq struct {
+	Log             telegraf.Logger `toml:"-"`
+	CPUs            []CPU           `toml:"-"`
+	PathSysfs       string          `toml:"host_sys"`
+	GatherThrottles bool            `toml:"gather_throttles"`
+}
+
+type CPU struct {
+	ID    string
+	Path  string
+	Props map[string]*os.File
+}
+
+var sampleConfig = `
+  ## Path for sysfs filesystem.
+  ## See https://www.kernel.org/doc/Documentation/filesystems/sysfs.txt
+  ## Defaults:
+  # host_sys = "/sys"
+
+  ## Gather CPU throttles per core
+  ## Defaults:
+  # gather_throttles = false
+`
+
+func (g *CPUFreq) SampleConfig() string {
+	return sampleConfig
+}
+
+func (g *CPUFreq) Description() string {
+	return "Collect current CPU's frequencies"
+}
+
+func (g *CPUFreq) Init() error {
+	if os.Getenv("HOST_SYS") != "" {
+		g.PathSysfs = os.Getenv("HOST_SYS")
+	} else {
+		g.PathSysfs = defaultHostSys
+	}
+
+	cpus, err := g.DiscoverCpus()
+	if err != nil {
+		return err
+	}
+	g.CPUs = cpus
+
+	return nil
+}
+
+func (g *CPUFreq) Gather(acc telegraf.Accumulator) error {
+	for _, cpu := range g.CPUs {
+		fields := make(map[string]interface{})
+		tags := map[string]string{"cpu": cpu.ID}
+
+		failed := false
+		for name, fd := range cpu.Props {
+			v, err := readUintFromFile(fd)
+			if err != nil {
+				acc.AddError(err)
+				failed = true
+				break
+			}
+
+			fields[name] = v
+		}
+
+		if !failed {
+			acc.AddFields("cpufreq", fields, tags)
+		}
+	}
+
+	return nil
+}
+
+func (g *CPUFreq) DiscoverCpus() ([]CPU, error) {
+	var cpus []CPU
+
+	cpuDirs, err := filepath.Glob(path.Join(g.PathSysfs, "devices/system/cpu/cpu[0-9]*"))
+	if err != nil {
+		return nil, err
```

Comment:
If `g.PathSysfs` contains path to nonexistent directory (because of rubbish in `HOST_SYS` env or in `host_sys` parameter in config file), `cpuDirs` will be empty forever. 

User will not get any metrics and will not get any information why these metrics are not here...

I would fail right here and inform user what is wrong. You can also check: `if len(cpuDirs) == 0` and return an error containing content of this command: `path.Join(g.PathSysfs, "devices/system/cpu/cpu[0-9]*")`